src/timely-util/src/builder_async.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Types to build async operators with general shapes.
+
+use std::cell::{Cell, RefCell};
+use std::future::Future;
+use std::mem::transmute;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll, Waker};
+
+use futures_util::task::ArcWake;
+
+use timely::communication::{message::RefOrMut, Pull};
+use timely::dataflow::channels::pact::ParallelizationContractCore;
+use timely::dataflow::channels::pushers::TeeCore;
+use timely::dataflow::channels::BundleCore;
+use timely::dataflow::operators::generic::builder_rc::OperatorBuilder as OperatorBuilderRc;
+use timely::dataflow::operators::generic::{InputHandleCore, OperatorInfo, OutputWrapper};
+use timely::dataflow::operators::{Capability, CapabilityRef};
+use timely::dataflow::{Scope, StreamCore};
+use timely::progress::{Antichain, Timestamp};
+use timely::scheduling::SyncActivator;
+use timely::{Container, PartialOrder};
+
+/// Builds async operators with generic shape.
+pub struct OperatorBuilder<G: Scope> {
+    builder: OperatorBuilderRc<G>,
+    /// Frontier information of input streams shared with the handles. Each frontier is paired with
+    /// a flag indicating whether or not the input handle has seen the updated frontier.
+    shared_frontiers: Rc<RefCell<Vec<(Antichain<G::Timestamp>, bool)>>>,
+    /// Wakers registered by input handles
+    registered_wakers: Rc<RefCell<Vec<Waker>>>,
+    /// The waker set up to activate this timely operator when woken
+    operator_waker: Arc<TimelyWaker>,
+    /// The number of active input handles
+    input_handles: Rc<Cell<usize>>,
+}
+
+/// An async Waker that activates a specific operator when woken and marks the task as ready
+struct TimelyWaker {
+    activator: SyncActivator,
+    active: AtomicBool,
+    task_ready: AtomicBool,
+}
+
+impl ArcWake for TimelyWaker {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        // Only activate the timely operator if it's not already active to avoid an infinite loop
+        if !arc_self.active.load(Ordering::SeqCst) {
```

Comment:
this can be `Relaxed` I think? we only need to order it with itself right?

Reply:
You also want to use `compact_exchange` instead of 2 load's and store's i think

Reply:
The load and store are on different flags. Where could I use `compare_exchange`?

Reply:
oh i misread it